src/mlpack/methods/logistic_regression/logistic_regression.hpp

```diff
@@ -37,79 +37,128 @@ template<typename MatType = arma::mat>
 class LogisticRegression
 {
  public:
+  typedef typename MatType::elem_type ElemType;
+  typedef typename GetDenseRowType<MatType>::type RowType;
+  typedef typename GetDenseColType<MatType>::type ColType;
+
+  /**
+   * Construct the LogisticRegression class without performing any training.
+   * The dimensionality of the data (which will be used to set the size of the
+   * parameters vector) must be specified, and all of the parameters in the
+   * model will be set to 0.  Note that the dimensionality may be changed later
+   * by directly modifying the parameters vector (using Parameters()).
+   *
+   * @param dimensionality Dimensionality of the data.
+   * @param lambda L2-regularization parameter.
+   */
+  LogisticRegression(const size_t dimensionality = 0,
+                     const double lambda = 0);
+
   /**
    * Construct the LogisticRegression class with the given labeled training
    * data.  This will train the model.  Optionally, specify lambda, which is the
    * penalty parameter for L2-regularization.  If not specified, it is set to 0,
    * which results in standard (unregularized) logistic regression.
    *
-   * It is not possible to set a custom optimizer with this constructor.  Either
-   * use a constructor that does not train and call Train() with a custom
-   * optimizer type, or use the constructor that takes an instantiated
-   * optimizer.  (This unfortunate situation is a language restriction of C++.)
+   * This constructor uses the default L-BFGS optimizer to train the model.
    *
    * @param predictors Input training variables.
    * @param responses Outputs resulting from input training variables.
    * @param lambda L2-regularization parameter.
    */
+  template<typename... CallbackTypes,
+           typename = typename std::enable_if<IsEnsCallbackTypes<
```

Comment:
Mostly I used the `IsEnsCallbackTypes` to avoid function ambiguity... it's not directly relevant for this one, but it is still nice to have.  For some of the `Train()` overloads, you could call e.g. `Train(data, responses, optimizer, 0.3)` or `Train(data, responses, optimizer, ens::TimerStop(0.1))` and there is no way to figure out if you meant that `0.3` should be `lambda` or a callback.  So the `IsEnsCallbackTypes` uses SFINAE to allow the compiler to figure out which overload is desired.